fix: #29

mrinalwahal opened the pull request:
Opening this PR as submission of my coding assignment.

This PR merges the following changes:

1. Fixes #29.
2.  Adds `ResolveFile()` and `ResolveFiles()` functions. 
3. Test function for `ResolveFile()` with a correct filename as input, a directory and a URL.

NOTE:

1. I've limited my solution strictly to the minimal description of issue #29.
1. I'm **NOT** very confident about either the utility of these 2 functions or my solutions. Here's why:
- I'm not sure why such simple functions were even required.
- Description on the issue page wasn't very explanatory.
- `ResolveFiles()` should ideally not exist. I'd say you should write a range to resolve multiple files whenever and wherever required. Its response is also not apt. There are better ways to return errors if any of those files can't be resolved. But I've kept the response strictly identical to the issue description.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/flanksource/commons?pullRequest=52) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/flanksource/commons?pullRequest=52) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/flanksource/commons?pullRequest=52) it.</sub>

moshloop commented on files/files.go:
```diff
@@ -316,14 +328,25 @@ func GetBaseName(filename string) string {
+func Getter(source, dst string) error {
+
+	//	Validate the URL first.
+	url, err := url.ParseRequestURI(source)
```
Why do we need to validate the URL first? Are we sure that url.ParseRequestURI will not return an error for all URLS/URI's supported by go-getter ?

mrinalwahal replied:
I think you are right. I added the URI parser primarily for branching the code to handle files and URLs separately. But then I forgot I shifted it to `ResolveFile()` function. I'm pushing the change.

moshloop commented on files/files.go:
```diff
@@ -359,3 +384,79 @@ func TempFileName(prefix, suffix string) string {
+	//
+	//	This will validate both a file in the filesystem,
+	//	as well as the URL source for a file.
+	if filepath.Ext(file) == "" {
```
Why does a file or URL need an extension? 

`$value == 1.00` is not a valid path or URL, yet would pass this test?

And likewise 
`/etc/hosts` is a valid path ?

mrinalwahal replied:
Aah, I only added the extension check to return an error if its a directory instead of a file. I totally forgot about files like `/etc/hosts` with no extensions. Let me improve this.

moshloop commented on files/files.go:
```diff
@@ -359,3 +384,79 @@ func TempFileName(prefix, suffix string) string {
+	//	Ensure the scheme is HTTP or HTTPS.
+	if err == nil &&
+		url.Scheme != "" &&
+		(strings.ToLower(url.Scheme) == "http" ||
```
What about the other schemes supported by go-getter ?

mrinalwahal replied:
I only added the scheme check because FS files came with `file://` scheme. And I wanted to exclude them. Because the `url.ParseRequestURI(source)` method is parsing even the URLs without a host as valid ones. So, I needed a quick way out.

mrinalwahal replied:
Oh, wait! I didn't know that `go-getter` also supports local files. I thought it only supported remote files over networks.

This calls for changing the entire function code now.

Apologies for this mishap! I'll get this done.

mrinalwahal replied:
@moshloop, should I add an extra argument for the destination to where we have to resolve the file? The issue description doesn't mention it, that's why I am asking. In absence of a valid destination, I'll initialize a temporary directory to resolve the files over there.

mrinalwahal replied:
Pushed new changes. Supports all `go-getter` schemes now 👍 Including `git://`.

moshloop commented:
@mrinalwahal can you check the test failures?  We can also drop support for 1.16

mrinalwahal commented:
> @mrinalwahal can you check the test failures? We can also drop support for 1.16

The test errors are related to `net/http` dependency not being located in GH Actions containers. I've updated the `go.mod` and `go.sum` files and dropped support for 1.13. Updated to 1.18.

Let me know if anything more is required.

Kaitou786 commented:
> > @mrinalwahal can you check the test failures? We can also drop support for 1.16
> 
> The test errors are related to `net/http` dependency not being located in GH Actions containers. I've updated the `go.mod` and `go.sum` files and dropped support for 1.13. Updated to 1.18.
> 
> Let me know if anything more is required.

Since you updated the go version in go.mod to 1.18 we won't have the go1.17 working with test... But we would need to support it so I suggest support from 1.17 and then fix the go.mod

mrinalwahal commented:
> > > @mrinalwahal can you check the test failures? We can also drop support for 1.16
> > 
> > 
> > The test errors are related to `net/http` dependency not being located in GH Actions containers. I've updated the `go.mod` and `go.sum` files and dropped support for 1.13. Updated to 1.18.
> > Let me know if anything more is required.
> 
> Since you updated the go version in go.mod to 1.18 we won't have the go1.17 working with test... But we would need to support it so I suggest support from 1.17 and then fix the go.mod

Hey @Kaitou786, I've also tried adding support from both 1.16 and 1.17 onward. But dependency management is breaking. Refreshed the `go.sum` and `go.mod` files as well for the same. Tests are still failing.

Also, I'd suggest from 1.16 onward because you seem to be running tests for it.

Kaitou786 commented:
> Also, I'd suggest from 1.16 onward because you seem to be running tests for it.

So would need to make pin to 1.16 on the go.mod?

> And the error is always coming from the same package i.e. `github.com/hairyhenderson/gomplate`. I think this package has issues in compiling with 1.16 onward.


you can try updating the GH setup/go action... I am able to compile your code locally with go1.17

mrinalwahal commented:
So the checks seem to have worked with support only from Go 1.18 onward.

But let me experiment once again with Go 1.17. If it doesn't work, we can switch to 1.18 again.

@Kaitou786 @moshloop

mrinalwahal commented:
Nope! Tried adding support for Go 1.17 and package `github.com/hairyhenderson/gomplate` once again threw errors.

So, I'm sticking with support from Go 1.18 onward only.
